Sources/NIO/ChannelPipeline.swift

```diff
@@ -944,7 +957,163 @@ extension ChannelPipeline {
+        }
+
+        /// Add one or more handlers to the pipeline.
+        ///
+        /// - Parameters:
+        ///   - handlers: The handlers to add.
+        ///   - position: The position in the `ChannelPipeline` to add `handlers`. Defaults to `.last`.
+        public func addHandlers(_ handlers: ChannelHandler...,
+                                position: ChannelPipeline.Position = .last) throws {
+            try self.pipeline.addHandlersSync(handlers, position: position).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the given handler instance if it is in
+        /// the `ChannelPipeline`, if it exists.
+        ///
+        /// - Parameter handler: The handler belonging to the context to fetch.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context(handler: ChannelHandler) throws -> ChannelHandlerContext {
+            return try self.pipeline._contextSync({ $0.handler === handler }).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the handler with the given name, if one exists.
+        ///
+        /// - Parameter name: The name of the handler whose context is being fetched.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context(name: String) throws -> ChannelHandlerContext {
+            return try self.pipeline.contextSync(name: name).get()
+        }
+
+        /// Returns the `ChannelHandlerContext` for the handler of given type, if one exists.
+        ///
+        /// - Parameter name: The name of the handler whose context is being fetched.
+        /// - Returns: The `ChannelHandlerContext` associated with the handler.
+        public func context<Handler: ChannelHandler>(handlerType: Handler.Type) throws -> ChannelHandlerContext {
+            return try self.pipeline._contextSync(handlerType: handlerType).get()
+        }
+
+        /// Returns the `ChannelHandler` of the given type from the `ChannelPipeline`, if it exists.
+        /// - Returns: A `ChannelHandler` of the given type if one exists in the `ChannelPipeline`.
+        public func handler<Handler: ChannelHandler>(type _: Handler.Type) throws -> Handler {
+            return try self.pipeline._handlerSync(type: Handler.self).get()
+        }
+    }
+
+    /// Invokes a closure with a synchronous view of this `ChannelPipeline` if this function is
+    /// called from the `EventLoop`.
+    ///
+    /// - Important: Users *must not* escape the `SynchronousView` from the closure.
+    /// - Parameter closure: A closure to operate on a synchronous view of this `ChannelPipeline`.
+    /// - Throws: `NIONotOnEventLoopError` if this function is not invoked on the event loop for
+    ///     this `ChannelPipeline`. Re-throws any error thrown from `body`.
+    public func withSynchronousViewIfAvailable(_ closure: (SynchronousView) throws -> Void) throws {
```

Comment:
- is all the extra syntax of the `with*` block worth it to make escaping the "SynchronousView" a little harder? I probably think it's not but can be convinced otherwise.
- I'd probably go for `channel.pipeline.syncOperations -> ChannelPipeline.SyncOperations` or something. Given that all operations throw _anyway_, why don't we just make them throw if you call the `SyncOperations` from off the EL? Then our interface is still thread-safe :). [problem: see below]

Reply:
I'm inclined to say that we should get rid of the ceremony of getting one of these, and then just force all of the methods to trap if they're used off-loop.

Reply:
The extra syntax is pretty minimal and -- I think -- makes it clearer that there are conditions to executing them synchronously (you have to think "can I really do this here?"). You also avoid having to check if you're on the event loop for each operation (and the caller only has to check whether we threw because we weren't on the event loop in one place). 

I'm not opposed `syncOperations` per se but I don't think it's any better.

Reply:
@glbrntt see my comment below, you can't actually check if it's safe or not (without trapping anyway).

Reply:
I saw, was just thinking about a solution! Seems like it's `syncOperations` and trapping :)

Reply:
@glbrntt or maybe `channel.pipeline.onEventLoop.addHandler` ?

Reply:
I prefer `syncOperations` but fine with either.